### PR TITLE
feat: setup a pyproject.toml

### PR DESCRIPTION
**Pull request by mistercrunch**
Decomposing a previous PR into smaller parts.

Using a pyproject.toml will enable us to use `supersetbot` to replace
dependabot in managing/bumping our python libraries. The fact that
pyproject is parsable enables more automation down the line

### CODE REVIEW

File: requirements/development.in [hunk 1]
```diff
@@ -17,26 +17,4 @@
...
-pyinstrument>=4.0.2,<5
-pylint
-pytest
-pytest-cov
-pytest-mock
-python-ldap>=3.4.4
-sqloxide
-statsd
-tox
+-e .[development,bigquery,cors,druid,gevent,gsheets,hive,mysql,playwright,postgres,presto,prophet,trino,thumbnails]
```
Comment (mistercrunch) [inline requirements/development.in:20]: note that I added `development` here, which is referencing a group that contains the libs referenced in the file bellow, now in pyproject.toml

**Review by eschutho — APPROVED** [review]
LGTM, but we may want more eyes on this.

**Comment by mistercrunch** [comment]
I don't think there's anything controversial here, just moving dep management to a new, parsable file that is becoming the standard. Mergin'

File: requirements/development.txt [hunk 2]
```diff
@@ -241,20 +244,15 @@ s3transfer==0.6.1
...
-    # via -r requirements/development.in
+    # via apache-superset
 tableschema==1.20.10
     # via apache-superset
-thrift==0.20.0
-    # via
-    #   apache-superset
-    #   pyhive
-    #   thrift-sasl
-thrift-sasl==0.4.3
```
Comment (john-bodley) [inline requirements/development.txt:256]: @mistercrunch the removal of `thrift-sasl` (which should be included via `pyhive[hive_pure_sasl]`) is breaking CI, i.e., for [test-postgres-hive](https://github.com/apache/superset/actions/runs/8532143694/job/23373369856?pr=27843#logs), I'm seeing:

```
> import thrift_sasl
E ModuleNotFoundError: No module named 'thrift_sasl'
```

Comment (mistercrunch) [reply requirements/development.txt:256]: I think that was bad merge conflict resolution, but interestingly this PR brought it back - guessing through a `pip-compile-multi --no-upgrade` run -> https://github.com/apache/superset/pull/27631/files#diff-f449835486933e5cab23a8567940e46d53edd89923cf1a3a4fb83005b069a93fR260

Now I did another run here -> https://github.com/apache/superset/pull/27861/

**Comment by john-bodley** [comment]
@mistercrunch is the intent to:

1. Remove `setup.py` and `setup.cfg`? I know we use the later for Mypy et al., but it seems that _most_ packages can use the `pyproject.toml` file for configuration ([example](https://mypy.readthedocs.io/en/stable/config_file.html))?
2. Update the [code and documentation](https://github.com/search?q=repo%3Aapache%2Fsuperset%20setup.py&type=code) to reference `pyproject.toml`?

**Comment by mistercrunch** [comment]
My main intent was to be able to parse the dependencies to support `supersetbot` acting as `dependabot` while moving to more modern means to store metadata. But I agree we should consolidate as much as possible, happy to follow up with a PR to try and kill setup.cfg is possible and update the docs where needed.

**Comment by mistercrunch** [comment]
https://github.com/apache/superset/pull/27884

**Pull request merged**